### PR TITLE
SAML 2.0 Assertion grant flow guide and concept

### PR DESCRIPTION
**Pull request by susanharper-okta**
## Description:
- **What's changed?** SAML 2.0 Assertion grant flow guide and concept
- **Is this PR related to a Monolith release?** Yes 2021-03-0

### Resolves:

* [OKTA-367012](https://oktainc.atlassian.net/browse/OKTA-367012)

### CODE REVIEW

File: packages/@okta/vuepress-site/docs/guides/implement-saml2/configure-policy-as/index.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+---
+title: Configure the Authorization Server policy
+---
+
+To enable consent for the [scopes](/docs/reference/api/authorization-servers/#create-a-scope) that you want to require consent for, select **API**, **Authorization Servers**, and then select **default** (Custom Authorization Server) in the table. In this example, we are enabling consent for default Custom Authorization Server scopes.
```
Comment (yuliu-okta) [inline packages/@okta/vuepress-site/docs/guides/implement-saml2/configure-policy-as/index.md:5]: Why enable consent?

File: packages/@okta/vuepress-site/docs/guides/implement-saml2/configure-policy-as/index.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+Make sure that the SAML 2.0 Assertion grant type is enabled in the Authorization Server default policy rule.
```
Comment (yuliu-okta) [inline packages/@okta/vuepress-site/docs/guides/implement-saml2/configure-policy-as/index.md:7]: it is not necessary to be in the default policy rule, the admin can either change an existing rule or create a new rule

File: packages/@okta/vuepress-site/docs/guides/implement-saml2/use-flow/index.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+If you are using the default Custom Authorization Server, then your request would look something like this:
+
+```bash
+curl --request POST \
+  --url https://${yourOktaDomain}/oauth2/default/v1/token \
+  --header 'accept: application/json' \
+  --header 'authorization: Basic MG9hDc....' \
+  --header 'content-type: application/x-www-form-urlencoded' \
+  --data 'grant_type=urn:ietf:params:oauth:grant-type:saml2-bearer' \
+  --data 'redirect_uri=https://example.com' \
```
Comment (yuliu-okta) [inline packages/@okta/vuepress-site/docs/guides/implement-saml2/use-flow/index.md:18]: it is not needed

File: packages/@okta/vuepress-site/docs/guides/implement-saml2/use-flow/index.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  --data 'assertion=<base64-encoded assertion>' \
+  --data 'scope=openid offline_access' \
+```
+
+> **Note:** The call to your authorization server's `/token` endpoint requires authentication. In this case, it is a Basic Auth digest of the Client ID and secret. You made note of these during <GuideLink link="../setup-app">app setup</GuideLink>. See [Client Authentication Methods](/docs/reference/api/oidc/#client-authentication-methods).
+
+Note the parameters that are being passed:
+
+- `grant_type`: `urn:ietf:params:oauth:grant-type:saml2-bearer`
+- `redirect_uri`: The callback location where the user agent is directed to along with the `access_token`. This must match one of the **Login redirect URIs** that you specified when you created your Okta application in the <GuideLink link="../setup-app">app setup</GuideLink> step.
```
Comment (yuliu-okta) [inline packages/@okta/vuepress-site/docs/guides/implement-saml2/use-flow/index.md:28]: it is not needed

File: packages/@okta/vuepress-site/docs/guides/implement-saml2/configure-policy-as/index.md (hunk 1 above)
Comment (susanharper-okta) [reply packages/@okta/vuepress-site/docs/guides/implement-saml2/configure-policy-as/index.md:5]: gah! I meant to delete that!!!!

File: packages/@okta/vuepress-site/docs/guides/implement-saml2/overview/index.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+At a high level, the SAML 2.0 Assertion flow has the following steps:
+
+- On behalf of a user, your web application attempts to access a resource that is hosted on a secure server.
+- The resource makes an authorization request to the SAML Identity Provider.
+- The SAML Identity Provider responds with a SAML 2.0 assertion for the already-authorized user.
+- The Okta client application requests an access token from the authorization server using the base64-encoded SAML 2.0 assertion as proof of identity.
+- The authorization server verifies the assertion and responds with an access token.
+- Your application extracts the token(s) from the HTTP response body.
+- Your application can now use the token(s) to call the resource server (for example, an API) on behalf of the user.
```
Comment (jakubvul-okta) [inline packages/@okta/vuepress-site/docs/guides/implement-saml2/overview/index.md:19]: Are the following the same thing, or different things?

"your web application"
"the Okta client application"
"Your application"

File: packages/@okta/vuepress-site/docs/guides/implement-saml2/setup-app/index.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+title: Set up your application
+---
+
+Create the client application that you want to use:
+
+1. From the Admin Console side navigation, select **Applications** and then **Applications**.
+
+2. On the Applications page, click **Add Application** and then **Create New App**.
+
+3. Select **Native** as the type of app **Platform** on the **Create a New Application Integration** window and then click **Create**.
```
Comment (jakubvul-okta) [inline packages/@okta/vuepress-site/docs/guides/implement-saml2/setup-app/index.md:11]:
```suggestion
3. Select **Native** as the type of app Platform on the **Create a New Application Integration** window and then click **Create**.
```

Comment (susanharper-okta) [reply packages/@okta/vuepress-site/docs/guides/implement-saml2/setup-app/index.md:11]: @jakubvul Platform is the name of the field in the dialog box. Do we not bold names of fields?

File: packages/@okta/vuepress-site/docs/concepts/oauth-openid/index.md [hunk 7]
```diff
@@ -272,3 +273,30 @@ app -> client: Response
 -->
 
 For information on how to set up your application to use this flow, see [Implement the Client Credentials Flow](/docs/guides/implement-client-creds/).
+
+### SAML 2.0 Assertion flow
+
+The SAML 2.0 Assertion flow is intended for a client app that wants to use an existing trust relationship without a direct user approval step at the authorization server. It enables a client application to obtain an existing authorization in the form of a valid, signed SAML assertion from a SAML Identity Provider. The client app can then exchange it for an OAuth access token from the OAuth authorization server. For example, this flow is useful when you want to fetch data from APIs that only support delegated permissions without prompting the user for credentials.
```
Comment (yuliu-okta) [inline packages/@okta/vuepress-site/docs/concepts/oauth-openid/index.md:279]: It does not `enables a client application to obtain an existing authorization`, should it be `It enables a client application to obtain an authorization from an existing trust relationship in the form of a valid, signed SAML assertion from a SAML Identity Provider. `

**Review by jakubvul-okta — APPROVED** [review]

**Review by yuliu-okta — APPROVED** [review]

**Pull request merged**